internal/lz4block/decode_arm64.s

```diff
@@ -0,0 +1,282 @@
+// +build gc
+// +build !noasm
+
+#include "go_asm.h"
+#include "textflag.h"
+
+// Register allocation.
+#define dst	R0
+#define dstorig	R1
+#define src	R2
+#define dstend	R3
+#define srcend	R4
+#define match	R5	// Match address.
+#define dictend	R6
+#define token	R7
+#define len	R8	// Literal and match lengths.
+#define offset	R7	// Match offset; overlaps with token.
+#define tmp1	R9
+#define tmp2	R11
+#define tmp3	R12
+
+// func decodeBlock(dst, src, dict []byte) int
+TEXT ·decodeBlock(SB), NOFRAME+NOSPLIT, $-8-80
+	MOVD dst_base  +0(FP), dst
+	MOVD dst_len   +8(FP), dstend
+	MOVD src_base +24(FP), src
+	MOVD src_len  +32(FP), srcend
+
+	CMP $0, srcend
+	BEQ shortSrc
+
+	ADD dst, dstend
+	ADD src, srcend
+
+	MOVD dst, dstorig
+
+loop:
+	// Read token. Extract literal length.
+	MOVBU.P 1(src), token
+	LSR     $4, token, len
+	CMP     $15, len
+	BNE     readLitlenDone
+
+readLitlenLoop:
+	CMP     src, srcend
+	BEQ     shortSrc
+	MOVBU.P 1(src), tmp1
+	ADDS    tmp1, len
+	BVS     shortDst
+	CMP     $255, tmp1
+	BEQ     readLitlenLoop
+
+readLitlenDone:
+	CMP $0, len
+	BEQ copyLiteralDone
+
+	// Bounds check dst+len and src+len.
+	ADDS     dst, len, tmp1
+	BCS      shortSrc
+	ADDS     src, len, tmp2
+	BCS      shortSrc
+	CMP      dstend, tmp1
+	BHI      shortDst
+	CMP      srcend, tmp2
+	BHI      shortSrc
+
+	// Copy literal.
+	CMP $8, len
+	BLO copyLiteralFinish
+
+	// Copy 0-7 bytes until src is aligned.
+	TST        $1, src
+	BEQ        twos
+	MOVBU.P    1(src), tmp1
+	MOVB.P     tmp1, 1(dst)
+	SUB        $1, len
+
+twos:
+	TST        $2, src
+	BEQ        fours
+	MOVHU.P    2(src), tmp2
+	MOVB.P     tmp2, 1(dst)
+	LSR        $8, tmp2, tmp1
+	MOVB.P     tmp1, 1(dst)
+	SUB        $2, len
+
+fours:
+	TST        $4, src
+	BEQ        copyLiteralLoopCond
+	MOVWU.P    4(src), tmp2
+	MOVB.P     tmp2, 1(dst)
+	LSR        $8, tmp2, tmp1
+	MOVB.P     tmp1, 1(dst)
+	LSR        $16, tmp2, tmp3
+	MOVB.P     tmp3, 1(dst)
+	LSR        $24, tmp2, tmp1
+	MOVB.P     tmp1, 1(dst)
+	SUB        $4, len
+
+	B copyLiteralLoopCond
+
+copyLiteralLoop:
+	// Aligned load, unaligned write.
```

Comment:
for 32-bit it wasn't worth the extra pain to do two aligned writes per register using masking vs the 4 from naive, but for 64-bit it might make a difference to go from 8 writes to 2 writes per 8 bytes.

thoughts @greatroar?

Reply:
My guess is that it's less expensive to do one unaligned read/write than many one-byte reads/writes. If you want to call memmove, you can reference the snappy code here: https://github.com/golang/snappy/blob/544b4180ac705b7605231d4a4550a1acb22a19fe/decode_arm64.s#L141

Reply:
ah, I see, so just write unaligned instead of trying to do fancy masking tricks. yeah, that's worth microbenchmarking.